src/dr_llm/llm/providers/response_validation.py

```
@@ -0,0 +1,73 @@
+    response_shape_error: str | None,
+) -> None:
+    """Raise the appropriate provider error for a non-OK HTTP response.
+
+    ``ProviderTransportError`` covers transient failures (5xx, 429, JSON decode
+    errors). ``ProviderSemanticError`` covers client errors and shape failures
+    that retrying will not resolve.
+    """
+    if status_code >= 500 or status_code == 429:
+        raise ProviderTransportError(
+            f"{provider_label} transient error status={status_code} body={response_text_preview}"
+        )
```

Ruff (0.15.9)

[warning] Avoid specifying long messages outside the exception class

(TRY003)